tree/ntuple/v7/src/RField.cxx

```diff
@@ -1751,6 +1757,126 @@ void ROOT::Experimental::RClassField::AcceptVisitor(Detail::RFieldVisitor &visit
+}
+
+ROOT::Experimental::RField<TObject>::RField(std::string_view fieldName)
+   : ROOT::Experimental::RFieldBase(fieldName, "TObject", ENTupleStructure::kRecord, false /* isSimple */)
+{
+#ifndef NDEBUG
+   static const auto cl = TObject::Class();
+#endif
+   assert(cl->GetClassVersion() == 1);
+   assert(cl->ClassProperty() & kClassHasExplicitCtor);
+   assert(cl->ClassProperty() & kClassHasExplicitDtor);
+   fTraits |= kTraitTriviallyConstructible | kTraitTriviallyDestructible;
+
+   Attach(std::make_unique<RField<UInt_t>>("fUniqueID"));
+   Attach(std::make_unique<RField<UInt_t>>("fBits"));
+}
+
+std::unique_ptr<ROOT::Experimental::RFieldBase>
+ROOT::Experimental::RField<TObject>::CloneImpl(std::string_view newName) const
+{
+   auto result = std::make_unique<RField<TObject>>(newName);
+   SyncFieldIDs(*this, *result);
+   return result;
+}
+
+std::size_t ROOT::Experimental::RField<TObject>::AppendImpl(const void *from)
+{
+   // Cf. TObject::Streamer()
+
+   auto *obj = static_cast<const TObject *>(from);
+   if (obj->TestBit(TObject::kIsReferenced)) {
+      throw RException(R__FAIL("RNTuple I/O on referenced TObject is unsupported"));
+   }
+
+   std::size_t nbytes = 0;
+   nbytes += CallAppendOn(*fSubFields[0], reinterpret_cast<const unsigned char *>(from) + GetOffsetUniqueID());
+
+   UInt_t bits = *reinterpret_cast<const UInt_t *>(reinterpret_cast<const unsigned char *>(from) + GetOffsetBits());
+   bits &= (~TObject::kIsOnHeap & ~TObject::kNotDeleted);
+   nbytes += CallAppendOn(*fSubFields[1], &bits);
+
+   return nbytes;
+}
+
+void ROOT::Experimental::RField<TObject>::ReadGlobalImpl(NTupleSize_t globalIndex, void *to)
+{
+   // Cf. TObject::Streamer()
+
+   auto *obj = static_cast<TObject *>(to);
+   if (obj->TestBit(TObject::kIsReferenced)) {
+      throw RException(R__FAIL("RNTuple I/O on referenced TObject is unsupported"));
+   }
+
+   CallReadOn(*fSubFields[0], globalIndex, static_cast<unsigned char *>(to) + GetOffsetUniqueID());
+
+   const UInt_t bitIsOnHeap = obj->TestBit(TObject::kIsOnHeap) ? TObject::kIsOnHeap : 0;
+   UInt_t bits;
+   CallReadOn(*fSubFields[1], globalIndex, &bits);
+   bits |= bitIsOnHeap | TObject::kNotDeleted;
+   *reinterpret_cast<UInt_t *>(reinterpret_cast<unsigned char *>(to) + GetOffsetBits()) = bits;
+}
+
+void ROOT::Experimental::RField<TObject>::OnConnectPageSource()
+{
+   if (GetTypeVersion() != 1) {
+      throw RException(R__FAIL("unsupported on-disk version of TObject: " + std::to_string(GetTypeVersion())));
+   }
+}
+
+std::uint32_t ROOT::Experimental::RField<TObject>::GetTypeVersion() const
+{
+   return TObject::Class()->GetClassVersion();
+}
+
+void ROOT::Experimental::RField<TObject>::ConstructValue(void *where) const
+{
+   new (where) TObject();
+}
+
+void ROOT::Experimental::RField<TObject>::RTObjectDeleter::operator()(void *objPtr, bool dtorOnly)
+{
+   static_cast<TObject *>(objPtr)->~TObject();
+   RDeleter::operator()(objPtr, dtorOnly);
```

Comment:
Is this more than:
```suggestion
   if (dtorOnly)
      static_cast<TObject *>(objPtr)->~TObject();
   else
      delete static_cast<TObject *>(objPtr);
```

>   static_cast<TObject *>(objPtr)->~TObject();
>   RDeleter::operator()(objPtr, dtorOnly);

This code (and its previous version) is confusing me ... if `dtorOnly == false` isn't the destructor called twice?

Reply:
What is the meaning of the 2nd parameter in:
```
   RDeleter::operator()(objPtr, dtorOnly);
```
if it is not to ask for the destructor to be called?

Reply:
If `dtorOnly` is true, the base class will do nothing. The alternative would be to have that if statement in all the derived classes.

Reply:
> If dtorOnly is true, the base class will do nothing.

This seems to be both by design and somewhat an happenstance (i.e. in a way the `RDeleter` base does call the destructor, it just happens that the destructor is always a no-op (actually doesn't exist) since it is for single instance of numerical types (no even arrays)).

Either way, it is 'hard' to read and maybe it would make more sense to use either the code I proposed or:
```
   static_cast<TObject *>(objPtr)->~TObject();
   RDeleter::operator()(objPtr, true /* dtorOnly */);
```
Related, I am not sure what is the value added of call `RDeleter` here rather than `operator delete`.

Reply:
Conclusion of off-line discussion: Use `RTypedDeleter` here to be consistent.  The usage of `RDeleter` here (and elsewhere) is to insure a single point of control for the way the user datum are allocated.